sonic-xcvrd/xcvrd/xcvrd.py

```diff
@@ -843,6 +846,376 @@ def is_fast_reboot_enabled():
+                        has_update = self.is_cmis_application_update_required(api, host_lanes, host_speed)
+                        if not has_update:
+                            # No application updates
+                            state = self.CMIS_STATE_READY
+                            self.log_notice("{}: state={}".format(lport, state))
+                            self.port_dict[lport]['cmis_state'] = state
+                            continue
+                        appl = self.get_cmis_application_desired(api, host_lanes, host_speed)
+                        self.port_dict[lport]['cmis_apsel'] = appl
+
+                        # D.2.2 Software Deinitialization
+                        api.set_datapath_deinit(host_lanes)
+                        api.set_lpmode(True)
+                        # D.1.3 Software Configuration and Initialization
+                        api.tx_disable_channel(host_lanes, True)
+                        api.set_lpmode(False)
+
+                        self.port_dict[lport]['cmis_state'] = self.CMIS_STATE_DP_DEINIT
+                    elif state == self.CMIS_STATE_DP_DEINIT:
+                        if api.get_module_state() != 'ModuleReady':
+                            continue
+
+                        # D.1.3 Software Configuration and Initialization
+                        api.set_application(host_lanes, self.port_dict[lport]['cmis_apsel'])
+
+                        self.port_dict[lport]['cmis_state'] = self.CMIS_STATE_AP_CONF
+                    elif state == self.CMIS_STATE_AP_CONF:
+                        st = api.get_config_datapath_hostlane_status()
+                        done = True
+                        for lane in range(self.NUM_CHANNELS):
+                            if ((1 << lane) & host_lanes) == 0:
+                                continue
+                            name = "ConfigStatusLane{}".format(lane + 1)
+                            if st[name] != 'ConfigSuccess':
+                                done = False
+                                continue
+                        if not done:
+                            continue
+
+                        # D.1.3 Software Configuration and Initialization
+                        api.set_datapath_init(host_lanes)
+
+                        self.port_dict[lport]['cmis_state'] = self.CMIS_STATE_DP_INIT
+                    elif state == self.CMIS_STATE_DP_INIT:
+                        st = api.get_datapath_state()
+                        done = True
+                        for lane in range(self.NUM_CHANNELS):
+                            if ((1 << lane) & host_lanes) == 0:
+                                continue
+                            name = "DP{}State".format(lane + 1)
+                            if st[name] != 'DataPathInitialized':
+                                done = False
+                                continue
+                        if not done:
+                            continue
```

Comment:
please bail out if datapath init failed. There is a timeout in page 1, byte 144

Reply:
This information is not available in the mem map as of now, hence we'll use a default expiration time for now, let's do the fine-grained expiration in phase 2 (DPB support)